docs/proposals/20191016-clusterctl-redesign.md

```diff
@@ -0,0 +1,271 @@
+- To manage the lifecycle of target, or workload, clusters.  This is the domain of Cluster API providers, and not clusterctl. (at this time).
+- To install network addons or any other components in the workload clusters.
+- To become a general purpose cluster management tool.  E.g. collect logs, monitor resources, etc.
+- To abstract away provider specific details.
+
+## Proposal
+
+In this section we will outline a high level overview of the proposed tool and its associated user stories.
+
+### Preconditions
+
+Prior to running clusterctl it is required that the operator has a valid KUBECONFIG + context to a running management cluster, following the same rules as kubectl.
+
+### User Stories
+
+#### Initial Deployment
+
+- As a Kubernetes operator, I’d like to have a simple way for clusterctl to install Cluster API providers into a management cluster, using a very limited set of resources and prerequisites, ideally from a laptop with something I can install with just a couple of terminal commands.
+- As a Kubernetes operator, I’d like to have the ability to provision clusters on different providers, using the same management cluster.
+- As a Kubernetes operator I would like to have a consistent user experience to configure and deploy clusters across cloud providers.
+
+#### Day Two Operations “Lifecycle Management”
+
+- As a Kubernetes operator I would like to be able to install new Cluster API providers into a management cluster.
+- As a Kubernetes operator I would like to be able to upgrade Cluster API components (including CRDs, controllers, etc).  
+- As a Kubernetes operator I would like to have a simple user experience to cleanly remove the Cluster API objects from a management cluster.
+
+#### Target Cluster Pivot Management
+
+- As a Kubernetes operator I would like to pivot Cluster API components from a management cluster to a target cluster.
+
+#### Provider Enablement
+
+- As a Cluster API provider developer, I would like to use my current implementation, or a new version, with clusterctl without recompiling.  
+
+### Implementation Details/Notes/Constraints
+
+> Portions of clusterctl may require changes to the Cluster API data model and/or the definition of new conventions for the providers in order to enable sets of user stories listed above, therefore a roadmap and feature deliverables will need to be coordinated over time.
+
+As of this writing, we envision clusterctl’s component model to look similar to the diagram below:
+
+![components](images/clusterctl-redesign/components.png)
+
+Clusterctl consists of a single binary artifact implementing a CLI; the same features exposed by the CLI will be made available to other management tools via a client library.
+
+During provider installation, clusterctl internal pkg/lib is expected to access provider repositories and read the yaml file specifying all the provider components to be installed in the management cluster.
```

Comment:
Maybe the internal library should be agnostic with respect of the source of this yaml and the cli or external tool should be the responsible of retrieving it. This would allow external tools to use other sources.

Reply:
clusterctl will let you specify alternative sources. Pulling from the provider repositories will be a default but not the only option. Does that address your comment?

Reply:
Yes I saw it below.

Reply:
@pablochacin just checking in - how are you feeling about this question?